File: original/game/src/ui/menuMap.ts
/**
 * The map behind the opening screen: the real Bay Area coastline with the real routes on it.
 *
 * Everything in this module is pure -- parse, project, frame -- so the parts that decide *where a
 * route appears* can be tested without a browser. The screen that draws it (`StartScreen`) owns the
 * DOM; this owns the arithmetic, because the arithmetic is where a map goes quietly wrong.
 *
 * The document is `game/public/menu-map.json`, written by `python -m sr.cli menu-map`. It carries
 * geometry and numbers only: names and blurbs live in the i18n files and the runtime reads them
 * from there (`docs/CONTRACT.md` section 6).
 */

export interface MenuRoute {
  id: string;
  km: number;
  checkpoints: number;
  /** [lon, lat] along the route, thinned by the pipeline to what a screen-wide map can resolve */
  line: [number, number][];
  /** Nearby OSM streets in the same lon/lat coordinates, so the enlarged route is a real map. */
  streets: { class: 'a' | 'b' | 'c'; line: [number, number][] }[];
}

/** [south, west, north, east] -- the window the pipeline cut the water and the land to. */
export type Region = [number, number, number, number];

export interface MenuMap {
  version: number;
  region: Region;
  /** closed rings: everything inside one of these is water */
  water: [number, number][][];
  /** closed rings: islands standing in that water */
  land: [number, number][][];
  /** open polylines: the motorways, trunks and primaries, drawn faintly so the land reads as land */
  roads: [number, number][][];
  /** the towns a local names when saying where something is; the label lives in i18n as `place.<id>` */
  places: { id: string; lon: number; lat: number }[];
  routes: MenuRoute[];
}

function bad(why: string): never {
  throw new Error(`menu map: ${why}`);
}

function line(value: unknown, where: string): [number, number][] {
  if (!Array.isArray(value) || value.length < 2) bad(`${where} is not a polyline`);
  return (value as unknown[]).map((p, i) => {
    if (!Array.isArray(p) || p.length !== 2 || !p.every((n) => typeof n === 'number' && Number.isFinite(n))) {
      bad(`${where}[${i}] is not a [lon, lat] pair`);
    }
    return [p[0] as number, p[1] as number];
  });
}

/** Read the document, or say why it cannot be read. Strict: half a map is worse than a list. */
export function parseMenuMap(data: unknown): MenuMap {
  if (typeof data !== 'object' || data === null) bad('not an object');
  const raw = data as Record<string, unknown>;
  if (raw.version !== 5) bad(`version ${String(raw.version)} is not 5`);
  if (!Array.isArray(raw.water) || !Array.isArray(raw.land)) bad('no water/land arrays');
  if (!Array.isArray(raw.roads)) bad('no roads array');
  if (!Array.isArray(raw.places)) bad('no places array');
  const places = raw.places.map((item, i) => {
    const p = item as Record<string, unknown>;
    if (typeof p.id !== 'string' || !p.id) bad(`places[${i}] has no id`);
    if (typeof p.lon !== 'number' || typeof p.lat !== 'number') bad(`places[${i}] has no position`);
    return { id: p.id, lon: p.lon, lat: p.lat };
  });
  const region = raw.region;
  if (!Array.isArray(region) || region.length !== 4
      || !region.every((n) => typeof n === 'number' && Number.isFinite(n))) bad('no region box');
  if (!Array.isArray(raw.routes) || raw.routes.length === 0) bad('no routes');
  const routes = raw.routes.map((item, i) => {
    const r = item as Record<string, unknown>;
    if (typeof r.id !== 'string' || !r.id) bad(`routes[${i}] has no id`);
    if (!Array.isArray(r.streets)) bad(`routes[${i}].streets is not an array`);
    const streets = r.streets.map((item, j) => {
      const street = item as Record<string, unknown>;
      const streetClass = street.class;
      if (streetClass !== 'a' && streetClass !== 'b' && streetClass !== 'c') {
        bad(`routes[${i}].streets[${j}] has bad class`);
      }
      const parsed: MenuRoute['streets'][number] = {
        class: streetClass, line: line(street.line, `routes[${i}].streets[${j}].line`),
      };
      return parsed;
    });
    return {
      id: r.id,
      km: typeof r.km === 'number' ? r.km : 0,
      checkpoints: typeof r.checkpoints === 'number' ? r.checkpoints : 0,
      line: line(r.line, `routes[${i}].line`),
      streets,
    } satisfies MenuRoute;
  });
  return {
    version: 5,
    region: region as Region,
    water: raw.water.map((c, i) => line(c, `water[${i}]`)),
    land: raw.land.map((c, i) => line(c, `land[${i}]`)),
    roads: raw.roads.map((c, i) => line(c, `roads[${i}]`)),
    places,
    routes,
  };
}

export interface Frame {
  west: number; east: number; south: number; north: number;
  /** how much a degree of longitude is worth beside a degree of latitude, at this frame's middle */
  kx: number;
}

/**
 * Trim the long axis of a frame until it matches the panel it is drawn in.
 *
 * Trimmed, not grown: the pipeline only cut water and land inside the region, so a frame grown past
 * it has a strip with no data in it, painted the land colour whatever is really there -- down the
 * Pacific side that was a black band beside the ocean.
 */
function fit(frame: Frame, aspect?: number): Frame {
  if (!aspect || aspect <= 0) return frame;
  let { west, east, south, north } = frame;
  const have = ((east - west) * frame.kx) / (north - south);
  if (aspect < have) {
    const trim = ((east - west) - (north - south) * aspect / frame.kx) / 2;
    west += trim; east -= trim;
  } else {
    const trim = ((north - south) - ((east - west) * frame.kx) / aspect) / 2;
    south += trim; north -= trim;
  }
  return { west, east, south, north, kx: frame.kx };
}

/**
 * The window the map is drawn through: the whole region the pipeline cut, never just the routes.
 *
 * A frame drawn around the routes answers "what does this route look like"; the player is
 * asking "where is it", and only the whole Bay answers that. The region stops at the Bay on purpose:
 * a route 150 km south would shrink the Bay into a smudge, so anything outside gets a marker on the
 * frame's edge instead -- which is what a real map does with an inset.
 */
export function frameOf(region: Region, aspect?: number): Frame {
  const [south, west, north, east] = region;
  const kx = Math.cos(((south + north) / 2) * Math.PI / 180);
  return fit({ west, east, south, north, kx }, aspect);
}

/**
 * The parts of a water ring that are actual shoreline, as open polylines.
 *
 * A water ring is closed by running along the edge of the region, so part of it is not coast at
 * all: drawing the ring's own outline would put a bright rectangle around the map. A segment whose
 * two ends both sit on the region's edge is that closure; everything else is shore.
 */
export function shoreLines(ring: [number, number][], region: Region,
                           eps = 1e-6): [number, number][][] {
  const [south, west, north, east] = region;
  const edge = (p: [number, number]): boolean =>
    Math.abs(p[0] - west) < eps || Math.abs(p[0] - east) < eps
    || Math.abs(p[1] - south) < eps || Math.abs(p[1] - north) < eps;
  const out: [number, number][][] = [];
  let cur: [number, number][] | null = null;
  for (let i = 0; i + 1 < ring.length; i++) {
    const a = ring[i]!, b = ring[i + 1]!;
    if (edge(a) && edge(b)) {
      if (cur) { out.push(cur); cur = null; }
      continue;
    }
    if (!cur) cur = [a];
    cur.push(b);
  }
  if (cur) out.push(cur);
  return out.filter((l) => l.length >= 2);
}

/** How tall the drawing is when it is `width` wide, so latitude and longitude stay in proportion. */
export function frameHeight(frame: Frame, width: number): number {
  return width * (frame.north - frame.south) / ((frame.east - frame.west) * frame.kx);
}

/**
 * A projector for one frame. Equirectangular with a cosine correction at the frame's middle: over
 * a bay sixty kilometres across the error is under a pixel, and a real projection would be a
 * dependency and a second coordinate system for nothing.
 */
export function projector(frame: Frame, width: number) {
  const height = frameHeight(frame, width);
  const spanX = (frame.east - frame.west) || 1e-9;
  const spanY = (frame.north - frame.south) || 1e-9;
  return (p: [number, number]): [number, number] => [
    (p[0] - frame.west) / spanX * width,
    (frame.north - p[1]) / spanY * height,
  ];
}

/** An SVG path for one polyline, at one decimal place -- more is bytes nobody can see. */
export function pathFor(poly: [number, number][], project: (p: [number, number]) => [number, number]): string {
  return poly.map((p, i) => {
    const [x, y] = project(p);
    return `${i ? 'L' : 'M'}${x.toFixed(1)} ${y.toFixed(1)}`;
  }).join('');
}

/* */
export function insideFrame(route: MenuRoute, frame: Frame): boolean {
  return route.line.some(([lon, lat]) =>
    lon >= frame.west && lon <= frame.east && lat >= frame.south && lat <= frame.north);
}

/** A route drawn on its own, in a box of its own: the shape you would recognise on a poster. */
export interface Circuit {
  /** an SVG path, in the box's coordinates */
  d: string;
  /** where the lap begins, for the marker that says so */
  start: [number, number];
  /** where it ends -- the same point on a loop, a different one on a point-to-point */
  end: [number, number];
  /** true when the route comes back to where it started, within a tenth of its own length */
  loop: boolean;
  /** Context paths projected by the same transform as `d`; never independently stretched. */
  streets: { class: 'a' | 'b' | 'c'; d: string }[];
}

/**
 * One route, fitted to a box, north up and in proportion.
 *
 * This is the circuit diagram, and it is the whole reason the opening screen reads as a racing game
 * rather than a map viewer. Every racing game ever made shows the lap this way, and the shape is already in the
 * document -- 140 points per route -- so it costs no new data at all.
 *
 * Longitude is scaled by cos(latitude) first: without it every route leans, because a degree east
 * is shorter than a degree north and nothing about the shape survives that.
 */
export function circuitPath(line: [number, number][], width: number, height: number,
                            pad = 14, streets: MenuRoute['streets'] = []): Circuit | null {
  if (line.length < 2) return null;
  const lat0 = line.reduce((s, p) => s + p[1], 0) / line.length;
  const kx = Math.cos(lat0 * Math.PI / 180);
  const xs = line.map((p) => p[0] * kx), ys = line.map((p) => p[1]);
  const west = Math.min(...xs), east = Math.max(...xs);
  const south = Math.min(...ys), north = Math.max(...ys);
  const spanX = (east - west) || 1e-9, spanY = (north - south) || 1e-9;
  const scale = Math.min((width - pad * 2) / spanX, (height - pad * 2) / spanY);
  const offX = (width - spanX * scale) / 2, offY = (height - spanY * scale) / 2;
  const at = (i: number): [number, number] =>
    [(xs[i]! - west) * scale + offX, (north - ys[i]!) * scale + offY];
  const project = (p: [number, number]): [number, number] =>
    [(p[0] * kx - west) * scale + offX, (north - p[1]) * scale + offY];
  let d = '';
  for (let i = 0; i < line.length; i++) {
    const [x, y] = at(i);
    d += `${i ? 'L' : 'M'}${x.toFixed(1)} ${y.toFixed(1)}`;
  }
  const start = at(0), end = at(line.length - 1);
  let run = 0;
  for (let i = 1; i < line.length; i++) {
    run += Math.hypot(xs[i]! - xs[i - 1]!, ys[i]! - ys[i - 1]!);
  }
  const gap = Math.hypot(xs[line.length - 1]! - xs[0]!, ys[line.length - 1]! - ys[0]!);
  return {
    d, start, end, loop: gap <= run * 0.1,
    streets: streets.map((street) => ({ class: street.class, d: pathFor(street.line, project) })),
  };
}

/** A rectangle in screen pixels, the way `getBoundingClientRect` gives it. */
export interface Box { left: number; top: number; right: number; bottom: number }

/**
 * The two lines of a magnifier callout: from a circle on the small map to the big drawing of it.
 *
 * This is the
 * convention every atlas and every exploded diagram uses -- a ring around the detail, two lines
 * fanning out to the enlargement -- and it works because the two lines *bracket* the big drawing:
 * they leave the ring at the extreme angles the target subtends, so the enlargement sits between
 * them rather than beside them.
 *
 * Returns null when the ring is inside the target or either shape has no size (a layout that has
 * not happened yet), because a callout with nothing to point at is a scribble.
 */
export function callout(cx: number, cy: number, r: number, box: Box):
    { x1: number; y1: number; x2: number; y2: number }[] | null {
  const w = box.right - box.left, h = box.bottom - box.top;
  if (!(w > 0) || !(h > 0) || !(r > 0)) return null;
  if (cx > box.left && cx < box.right && cy > box.top && cy < box.bottom) return null;
  const corners: [number, number][] = [
    [box.left, box.top], [box.right, box.top], [box.right, box.bottom], [box.left, box.bottom],
  ];
  // The widest pair as seen from the ring. Angles are measured from the ring's centre and compared
  // by their difference, so a fan that straddles the -pi/pi seam is still measured correctly.
  let best: [number, number] | null = null, widest = -1;
  for (let i = 0; i < corners.length; i++) {
    for (let j = i + 1; j < corners.length; j++) {
      const a = Math.atan2(corners[i]![1] - cy, corners[i]![0] - cx);
      const b = Math.atan2(corners[j]![1] - cy, corners[j]![0] - cx);
      let d = Math.abs(a - b);
      if (d > Math.PI) d = 2 * Math.PI - d;
      if (d > widest) { widest = d; best = [i, j]; }
    }
  }
  if (!best) return null;
  return best.map((i) => {
    const [x, y] = corners[i]!;
    const a = Math.atan2(y - cy, x - cx);
    return { x1: cx + Math.cos(a) * r, y1: cy + Math.sin(a) * r, x2: x, y2: y };
  });
}

File: world-tour/pipeline/sr/menumap.py
"""The one map the player sees before driving: the world, with every city's route on it.

The opening screen is a map, not a list, because the promise of this game is 「路是真的」.
For a world tour the map that says where a route is, is the world: the continents drawn from the
same elevation data every track stands on (land is where the ground reads above sea level), a pin
and a name for each city, and the routes themselves -- too small to see at this scale, so the screen
rings the chosen one and draws it big beside the map.

Output is `game/public/menu-map.json`, in lat/lon. The runtime projects; a menu that has to be
re-generated because the panel resized would be the wrong split.
"""
from __future__ import annotations

import glob
import gzip
import json
import math
import os

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
OUT = os.path.join(ROOT, "game", "public", "menu-map.json")

# The window the menu map is cut to, as (south, west, north, east): the world with its seam in the
# Pacific, where no route is. Wide enough east for Sydney and west for New York and Machu Picchu, and
# tall enough that a square drawing of it is mostly map rather than empty band.
REGION = (-60.0, -100.0, 78.0, 170.0)
# Land from the Terrarium elevation tiles at zoom 3 (about 20 km a pixel), sampled every quarter
# degree. The Caspian and the Dead Sea read below zero and come out as water, which they are; so does
# the Black Sea, whose bathymetry the tiles carry.
LAND_ZOOM = 3
LAND_STEP = 0.25
# A quarter-degree grid drawn 1000 units across 270 degrees: under a unit per cell, so the outline is
# simplified to half a degree and islands under two square degrees (a few pixels) are dropped.
LAND_TOLERANCE = 0.3
MIN_LAND_AREA = 2.0
LAND_CACHE = os.path.join(ROOT, "pipeline", "cache", "world", "land.json.gz")
# A route drawn on a screen-wide map needs its shape, not its samples: the pipeline stores a point
# every two metres, and 140 of them is more than a 900-pixel-wide map can resolve.
ROUTE_POINTS = 140
# An island simplifies to fewer than this many points when it is a rock, not a place.
MIN_ISLAND_POINTS = 5


def fetch_land(force=False, log=print):
    """The continents as rings, from the elevation tiles, cached like every other layer."""
    if os.path.exists(LAND_CACHE) and not force:
        return LAND_CACHE
    import numpy as np
    from shapely.geometry import box
    from shapely.ops import unary_union
    from sr.dem import DemSampler
    from sr.fetch_dem import fetch_tile
    s, w, n, e = REGION
    tiles = 2 ** LAND_ZOOM
    for x in range(tiles):
        for y in range(tiles):
            fetch_tile(x, y, LAND_ZOOM, log=lambda *_: None)
    dem = DemSampler(LAND_ZOOM, fallback_zoom=None)
    lats = np.arange(s + LAND_STEP / 2, n, LAND_STEP)
    lons = np.arange(w + LAND_STEP / 2, e, LAND_STEP)
    cells = []
    for lat in lats:
        land = dem.heights(np.full(len(lons), lat), lons) > 0
        # One rectangle per run of land cells in a row: a few thousand boxes instead of 400 thousand.
        edges = np.flatnonzero(np.diff(np.r_[0, land.astype(int), 0]))
        for i0, i1 in zip(edges[::2], edges[1::2]):
            cells.append(box(w + i0 * LAND_STEP, lat - LAND_STEP / 2, w + i1 * LAND_STEP, lat + LAND_STEP / 2))
    # Simplified by more than a cell, so the outline is a coast and not the grid's staircase.
    merged = unary_union(cells).simplify(LAND_STEP * 1.2)
    polys = [merged] if merged.geom_type == "Polygon" else list(merged.geoms)
    rnd = lambda coords: [[round(x, 3), round(y, 3)] for x, y in coords]
    from shapely.geometry import Polygon
    rings = [rnd(poly.exterior.coords) for poly in polys if poly.area >= MIN_LAND_AREA]
    # The seas a continent closes round -- the Mediterranean, the Black Sea, the Caspian, Hudson Bay --
    # are holes in its outline. Dropping them paints the Mediterranean as desert.
    seas = [rnd(hole.coords) for poly in polys if poly.area >= MIN_LAND_AREA
            for hole in poly.interiors if Polygon(hole).area >= MIN_LAND_AREA]
    os.makedirs(os.path.dirname(LAND_CACHE), exist_ok=True)
    with gzip.open(LAND_CACHE, "wt", encoding="utf-8") as fh:
        json.dump({"version": 2, "region": list(REGION), "zoom": LAND_ZOOM, "step": LAND_STEP,
                   "rings": rings, "seas": seas}, fh)
    log(f"land: {len(rings)} rings, {len(seas)} enclosed seas, {sum(len(r) for r in rings + seas)} points")
    return LAND_CACHE


def land_rings(path=None, key="rings"):
    """The cached continents (`rings`) or the seas they close round (`seas`)."""
    path = path or LAND_CACHE
    if not os.path.exists(path):
        return []
    with gzip.open(path, "rt", encoding="utf-8") as fh:
        return [[tuple(p) for p in ring] for ring in json.load(fh).get(key, [])]


def simplify(points, tolerance):
    """Douglas-Peucker. Keeps the shape of a shoreline; a plain stride would eat its headlands."""
    if len(points) < 3:
        return list(points)
    (x0, y0), (x1, y1) = points[0], points[-1]
    dx, dy = x1 - x0, y1 - y0
    span = math.hypot(dx, dy) or 1e-12
    worst, at = 0.0, 0
    for i in range(1, len(points) - 1):
        x, y = points[i]
        d = abs(dy * x - dx * y + x1 * y0 - y1 * x0) / span
        if d > worst:
            worst, at = d, i
    if worst <= tolerance:
        return [points[0], points[-1]]
    return simplify(points[:at + 1], tolerance)[:-1] + simplify(points[at:], tolerance)


def simplify_ring(ring, tolerance):
    """Douglas-Peucker for a closed ring.

    Straight to `simplify` a ring collapses to nothing: its first and last point are the same, so
    the base line every distance is measured against has zero length and every distance comes out
    zero. Cutting the ring at its farthest point first gives two open lines with real ends.
    """
    pts = ring[:-1] if len(ring) > 1 and ring[0] == ring[-1] else list(ring)
    if len(pts) < 4:
        return list(ring)
    x0, y0 = pts[0]
    far = max(range(1, len(pts)), key=lambda i: (pts[i][0] - x0) ** 2 + (pts[i][1] - y0) ** 2)
    head = simplify(pts[:far + 1], tolerance)
    tail = simplify(pts[far:] + [pts[0]], tolerance)
    return head[:-1] + tail


def chains(ways):
    """Join coastline ways end to end, forwards and backwards.

    OSM stores a shoreline as thousands of separate ways, and the direction of each one carries the
    only fact that matters here: **land is on the left**. Joining keeps that fact and turns the
    pieces into a handful of long shorelines. Joining in one direction only is not enough -- start
    from a way in the middle and the shoreline splits in two, and the half that ends *inside* the
    frame has no boundary crossing to close its polygon against.
    """
    by_start, by_end = {}, {}
    for w in ways:
        by_start.setdefault(w[0], []).append(w)
        by_end.setdefault(w[-1], []).append(w)
    used, out = set(), []
    for w in ways:
        if id(w) in used:
            continue
        used.add(id(w))
        chain = list(w)
        while chain[-1] != chain[0]:
            nxt = next((c for c in by_start.get(chain[-1], []) if id(c) not in used), None)
            if nxt is None:
                break
            used.add(id(nxt)); chain.extend(nxt[1:])
        while chain[-1] != chain[0]:
            prev = next((c for c in by_end.get(chain[0], []) if id(c) not in used), None)
            if prev is None:
                break
            used.add(id(prev)); chain[:0] = prev[:-1]
        out.append(chain)
    return out


def _clip_segment(p, q, rect):
    """Liang-Barsky: the part of segment p->q inside the rect, as (t0, t1), or None."""
    s, w, n, e = rect
    dx, dy = q[0] - p[0], q[1] - p[1]
    t0, t1 = 0.0, 1.0
    for num, den in ((p[0] - w, -dx), (e - p[0], dx), (p[1] - s, -dy), (n - p[1], dy)):
        if den == 0:
            if num < 0:
                return None
        else:
            r = num / den
            if den < 0:
                if r > t1:
                    return None
                t0 = max(t0, r)
            else:
                if r < t0:
                    return None
                t1 = min(t1, r)
    return (t0, t1) if t1 > t0 else None


def clip_chain(chain, rect):
    """A chain cut to the rect: a list of polylines, each entering and leaving at the boundary."""
    pieces, cur = [], None
    for p, q in zip(chain, chain[1:]):
        got = _clip_segment(p, q, rect)
        if got is None:
            if cur:
                pieces.append(cur); cur = None
            continue
        t0, t1 = got
        dx, dy = q[0] - p[0], q[1] - p[1]
        a = (p[0] + dx * t0, p[1] + dy * t0)
        b = (p[0] + dx * t1, p[1] + dy * t1)
        if cur is None or t0 > 0:
            if cur:
                pieces.append(cur)
            cur = [a]
        if cur[-1] != b:
            cur.append(b)
        if t1 < 1:
            pieces.append(cur); cur = None
    if cur:
        pieces.append(cur)
    return [p for p in pieces if len(p) >= 2]


def _edge_t(pt, rect, eps=1e-9):
    """Where a boundary point sits on the rect, as a number in [0, 4) going *clockwise* from the
    north-west corner. Clockwise is the whole trick: land is on the left of a coastline, so water is
    on its right, and a ring that follows the shore and then turns clockwise along the frame keeps
    the water inside it."""
    s, w, n, e = rect
    x, y = pt
    wide, tall = (e - w) or 1e-12, (n - s) or 1e-12
    if abs(y - n) <= eps:
        return (x - w) / wide
    if abs(x - e) <= eps:
        return 1 + (n - y) / tall
    if abs(y - s) <= eps:
        return 2 + (e - x) / wide
    if abs(x - w) <= eps:
        return 3 + (y - s) / tall
    return None


def _corners(rect):
    s, w, n, e = rect
    return [(0.0, (w, n)), (1.0, (e, n)), (2.0, (e, s)), (3.0, (w, s))]


def water_rings(chains_in, rect, eps=1e-7):
    """Every stretch of water inside the rect, as closed rings, plus the islands sitting in it.

    Returns (water, islands). The water rings are assembled from the clipped shoreline: follow a
    piece from where it enters the frame to where it leaves, then run clockwise along the frame
    edge to wherever the next piece enters, and keep going until the ring closes. Islands are the
    coastline ways that close on themselves -- they are land, and get drawn on top of the water.
    """
    pieces, islands = [], []
    for chain in chains_in:
        closed = chain[0] == chain[-1]
        for piece in clip_chain(chain, rect):
            a, b = _edge_t(piece[0], rect, eps), _edge_t(piece[-1], rect, eps)
            if a is None or b is None:
                if closed and piece[0] == piece[-1]:
                    islands.append(piece)      # an island wholly inside the frame
                continue                       # a fragment with a loose end: nothing to close it
            pieces.append((a, b, piece))
    pieces.sort(key=lambda x: x[0])
    corners = _corners(rect)
    water, done = [], set()
    for i in range(len(pieces)):
        if i in done:
            continue
        ring, at = [], i
        while at not in done:
            done.add(at)
            _, out_t, poly = pieces[at]
            ring.extend(poly)
            nxt = min(range(len(pieces)),
                      key=lambda j: ((pieces[j][0] - out_t) % 4.0) or 4.0)
            in_t = pieces[nxt][0]
            span = (in_t - out_t) % 4.0
            for ct, pt in corners:
                if 0 < (ct - out_t) % 4.0 <= span:
                    ring.append(pt)
            at = nxt
        if len(ring) >= 4:
            ring.append(ring[0])
            water.append(ring)
    return water, islands


def local_line(track, points, limit=None):
    """Local x/z metre pairs as lat/lon, optionally thinned to `limit` points."""
    lat0, lon0 = track["origin"]["lat"], track["origin"]["lon"]
    k = math.cos(math.radians(lat0))
    stride = max(1, len(points) // limit) if limit else 1
    return [[round(lon0 + p[0] / (111320.0 * k), 5), round(lat0 - p[1] / 111320.0, 5)]
            for p in points[::stride]]


def route_line(track, points=ROUTE_POINTS):
    """A track's spline as lat/lon, thinned to `points`.

    The spline is in local metres about the track's own origin, so each track converts with its own
    latitude scale -- one shared scale would shear Monterey against Berkeley by a whole street.
    """
    return local_line(track, [[p[0], p[2]] for p in track["spline"]["points"]], points)


def nearby_streets(track, map_path, margin=0.3):
    """Roads intersecting the route's padded frame, in the exact same coordinates as the route.

    The per-track street map already contains all OSM streets around the drive. Keeping only roads
    near the route stops a wide OSM fetch rectangle from shrinking the selected circuit into a dot;
    the UI clips the remaining overshoot at the frame edge.
    """
    if not os.path.exists(map_path):
        return []
    with open(map_path, encoding="utf-8") as fh:
        doc = json.load(fh)
    route = [[p[0], p[2]] for p in track["spline"]["points"]]
    xs, zs = [p[0] for p in route], [p[1] for p in route]
    dx, dz = max(xs) - min(xs), max(zs) - min(zs)
    px, pz = max(100.0, dx * margin), max(100.0, dz * margin)
    frame = (min(xs) - px, min(zs) - pz, max(xs) + px, max(zs) + pz)
    out = []
    for road in doc.get("roads", []):
        flat = road.get("p", [])
        line = [[flat[i], flat[i + 1]] for i in range(0, len(flat) - 1, 2)]
        if len(line) < 2:
            continue
        rx, rz = [p[0] for p in line], [p[1] for p in line]
        if max(rx) < frame[0] or min(rx) > frame[2] or max(rz) < frame[1] or min(rz) > frame[3]:
            continue
        out.append({"class": road.get("c", "c"), "line": local_line(track, line)})
    return out


def build(tracks_root=None, cache=None, tolerance=LAND_TOLERANCE):
    """The whole document: the sea as the ground, the continents on it, every city and its route.

    Names and blurbs are deliberately absent -- they live in the i18n files and the runtime reads
    them from there (docs/CONTRACT.md section 6).
    """
    rings = land_rings(cache)
    if not rings:
        raise RuntimeError(f"no land at {cache or LAND_CACHE}; "
                           "run: .venv/bin/python -c 'from sr import menumap; menumap.fetch_land()'")
    s, w, n, e = REGION
    land = [r for r in (simplify_ring(r, tolerance) for r in rings) if len(r) >= MIN_ISLAND_POINTS]
    seas = [r for r in (simplify_ring(r, tolerance) for r in land_rings(cache, "seas")) if len(r) >= MIN_ISLAND_POINTS]
    root = tracks_root or os.path.join(ROOT, "game", "public", "tracks")
    routes, pins = [], []
    for path in sorted(glob.glob(os.path.join(root, "*", "track.json"))):
        with open(path, encoding="utf-8") as fh:
            track = json.load(fh)
        if track["id"].startswith("synth"):
            continue                       # test fixtures, not places
        routes.append({
            "id": track["id"],
            "category": track.get("category"),
            # A loop's spline is one physical lap, while this number is what the player is signing up to drive.
            "km": round(track["spline"]["length"] * int(track.get("laps", 1)) / 1000.0, 2),
            "checkpoints": len(track.get("checkpoints", [])),
            "line": route_line(track),
            "streets": nearby_streets(track, os.path.join(os.path.dirname(path), "map.json")),
        })
        # One pin per city, at its track's origin; the label is the i18n key `place.<track id>`.
        pins.append({"id": track["id"], "lon": round(track["origin"]["lon"], 4), "lat": round(track["origin"]["lat"], 4)})
    rnd = lambda ring: [[round(x, 3), round(y, 3)] for x, y in ring]
    # A world map is sea with land on it: the ground is water, the continents are drawn on it, and
    # `water` is the enclosed seas drawn back over the land (the runtime paints in that order).
    return {"version": 5, "world": True, "region": list(REGION),
            "water": [rnd(r) for r in seas],
            "land": [rnd(r) for r in land], "roads": [],
            "places": pins, "routes": routes}


def write(out_path=None, tracks_root=None, cache=None):
    doc = build(tracks_root, cache)
    path = out_path or OUT
    os.makedirs(os.path.dirname(path), exist_ok=True)
    with open(path, "w", encoding="utf-8") as fh:
        json.dump(doc, fh, ensure_ascii=False, separators=(",", ":"))
    return doc
